ide/project.dependency/src/org/netbeans/modules/project/dependency/ArtifactSpec.java

```diff
@@ -0,0 +1,193 @@
+    }
+
+    @Override
+    public int hashCode() {
+        int hash = 7;
+        hash = 79 * hash + Objects.hashCode(this.kind);
+        hash = 79 * hash + Objects.hashCode(this.type);
+        hash = 79 * hash + Objects.hashCode(this.groupId);
+        hash = 79 * hash + Objects.hashCode(this.artifactId);
+        hash = 79 * hash + Objects.hashCode(this.versionSpec);
+        hash = 79 * hash + Objects.hashCode(this.classifier);
+        return hash;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null) {
+            return false;
+        }
+        if (getClass() != obj.getClass()) {
+            return false;
+        }
+        final ArtifactSpec<?> other = (ArtifactSpec<?>) obj;
+        if (!Objects.equals(this.type, other.type)) {
+            return false;
+        }
+        if (!Objects.equals(this.groupId, other.groupId)) {
+            return false;
+        }
+        if (!Objects.equals(this.artifactId, other.artifactId)) {
+            return false;
+        }
+        if (!Objects.equals(this.versionSpec, other.versionSpec)) {
+            return false;
+        }
+        if (!Objects.equals(this.classifier, other.classifier)) {
+            return false;
+        }
+        return this.kind == other.kind;
+    }
+    
+    public String toString() {
+        StringBuilder sb = new StringBuilder(
+            String.format("%s:%s:%s", getGroupId(), getArtifactId(), getVersionSpec() == null ? "" : getVersionSpec())
+        );
+        if (classifier != null) {
+            sb.append(":").append(classifier);
+        }
+        if (type != null) {
+            sb.append("[").append(type).append("]");
+        }
+        if (optional) {
+            sb.append("?");
+        }
+        return sb.toString();
+    }
+    
+    /**
+     * Returns opaque project-specific data. If searching for
+     * a project-specific extension, use {@link ProjectDependencies#findAdapters} instead.
+     * 
+     * @return unspecified underlying project data
+     */
+    public T getProjectData() {
+        return data;
+    }
+    
+    public static <V> ArtifactSpec<V> createVersionSpec(
+            @NonNull String groupId, @NonNull String artifactId, 
+            @NullAllowed String type, @NullAllowed String classifier, 
+            @NonNull String versionSpec, boolean optional, @NonNull V data) {
+        return new ArtifactSpec<V>(VersionKind.REGULAR, groupId, artifactId, versionSpec, type, classifier, optional, data);
+    }
+
+    public static <V> ArtifactSpec<V> createSnapshotSpec(
+            @NonNull String groupId, @NonNull String artifactId, 
+            @NullAllowed String type, @NullAllowed String classifier, 
+            @NonNull String versionSpec, boolean optional, @NonNull V data) {
+        return new ArtifactSpec<V>(VersionKind.SNAPSHOT, groupId, artifactId, versionSpec, type, classifier, false, data);
+    }
+
```

Comment:
And VersionKind.Latest? We don't need it now?

Reply:
Removed in 0221749c9a